src/server/tiered_storage.cc

```diff
@@ -454,55 +516,57 @@ void TieredStorage::WriteSingle(DbIndex db_index, PrimeIterator it, size_t blob_
     return;
   }
 
-  constexpr size_t kMask = kBlockAlignment - 1;
-  size_t page_size = (blob_len + kMask) & (~kMask);
-
-  DCHECK_GE(page_size, blob_len);
-  DCHECK_EQ(0u, page_size % kBlockAlignment);
+  SingleRequest* req = new SingleRequest(blob_len, res, it->first.ToString());
 
-  struct SingleRequest {
-    char* block_ptr = nullptr;
-    PrimeTable* pt = nullptr;
-    size_t blob_len = 0;
-    off_t offset = 0;
-    string key;
-  } req;
+  auto& enqueued_entries = db_arr_[db_index]->bigbin_enqueued_entries;
+  auto emplace_res = enqueued_entries.emplace(req->key, req);
+  CHECK(emplace_res.second);
 
-  char* block_ptr = (char*)mi_malloc_aligned(page_size, kBlockAlignment);
-
-  req.blob_len = blob_len;
-  req.offset = res;
-  req.key = it->first.ToString();
-  req.pt = db_slice_.GetTables(db_index).first;
-  req.block_ptr = block_ptr;
-
-  it->second.GetString(block_ptr);
+  it->second.GetString(req->block_ptr);
   it->second.SetIoPending(true);
 
-  auto cb = [req = std::move(req)](int io_res) {
-    PrimeIterator it = req.pt->Find(req.key);
-    CHECK(!it.is_done());
+  auto cb = [this, req, db_index](int io_res) {
+    PrimeTable* pt = db_slice_.GetTables(db_index).first;
+    PrimeIterator it = pt->Find(req->key);
+    absl::Cleanup cleanup = [this, req]() {
+      mi_free(req->block_ptr);
+      delete req;
+      --num_active_requests_;
+    };
+    // In case entry was deleted or request was canceled free allocated.
+    if (it.is_done() || req->cancel) {
```

Comment:
i suggest checking `req->cancel` before calling find.
is there a case where `it.is_done()` is true when req->cancel is false ?

Reply:
you are write, we should not get to it.is_done() if req->cancel is true.
I will fix this